### ISSUE TITLE
[RFC] Future of the Symfony HttpCache and its usage in production 

### ISSUE THREAD

**Issue by tgalopin** [symfony/symfony#19705]
Hello everyone,

I open this issue as a request for debate around the Symfony HttpCache and especially about its usage in development and production.

A few days ago, I stumbled upon the issue https://github.com/symfony/symfony/issues/4871 (HttpCache files getting too big) which is a problem only because the HttpCache is used in production.

Philosophically, the Symfony HttpCache was designed as both a development tool to get logs about the application HTTP cache configuration (hits, misses, etc.) and as in production environments as a simpler replacement for Varnish or other HTTP proxies (especially when the application run on a shared host).

While this usage in production is currently supported (but not recommended), it has limitations and issues (see https://github.com/symfony/symfony/issues?utf8=%E2%9C%93&q=is%3Aopen%20is%3Aissue%20httpcache%20). I would like to discuss the evolution possibilities in this RFC.

I see four options (but there are probably more, don't hesitate to propose):

1. We create two separated modes: one for dev and one for prod. The one for dev is integrated into the debug tools and the other one is designed for production.

   Pros: current installations can be migrated to the production version while having the debug tools for HTTP cache in the profiler.

   Cons: We have to deal with the usage of HttpCache in production officially, which requires maintenance and support. We also have to fix the current issues with HttpCache in production.

2. We don't change the code and simply explain in the documentation that the HttpCache is not designed for production. We could also add documentation about how to use it in development.

   Pros: easiy to implement, does not require code changes.

   Cons: I think it introduces more confusion than solves anything and is likely to don't solve the issues we currently have.

3. We don't change the code and keep the support of production but we list in documentation the known limitations and officially explain these won't be fixed.

   Pros: easiy to implement, does not require code changes and is clearer than 2.

   Cons: supporting HttpCache officially in production with limitations may become a problem to some people.

4. We deprecate the usage of the HttpCache in production in 3.2/3.3 and remove the possibility in 4.0. In parallel, we integrate it more deeply with the debug profiler in order to get the same debug informations (hits/misses/...) but in a debug environment (and therefore we become clear about the fact that the HttpCache should only be used in development). We could provide this HTTP cache debug features as an opt-in in 3.X and enabled by default in 4.0.

   Pros: it shows the idea behind the Symfony HttpCache and answer the issues with its usage in production (don't use it in production).

   Cons: current people having to use the HttpCache in production as they don't control their environment (shared environments for instance) won't be able to use an alternative and won't get any HTTP cache anymore.

I would like to get opinions. What do you think? Do you have an interesting use-case for this debate?

**Comment by stof** [comment]
Integrating the HttpCache stats in the debug profiler might be hard, because the debug profiler is part of the app code running only on cache miss. What kind of deeper integration are you thinking about ?

**Comment by tgalopin** [comment]
I don't have a specific implementaton in mind. What I propose is to clearly define HttpCache as a debug tool, by including it in the debug tools (somehow) and by removing its usage in the app.php file (in production).

**Comment by jpauli** [comment]
I think that sitting some PHP code, in front of a PHP designed application, just to prevent running it; is really silly.

PHP's request startup / shutdown time is way slower than a dedicated tool such as Varnish (which has been designed for that, so its source code has been oriented with just one goal, aka it got hugepage support, sendfile() support, IPC support and everything needed for such a task).

PHP will be slower, less scalable and configurable than a dedicated tool, and I find that investing some developpers time to create something that already exists and has been proven to work, is silly to me.

**Comment by vschoener** [comment]
@tgalopin May be it will be better to push them using a real Proxy Cache as Varnish ? 

HttpCache is great to understand the usage and how things work, but to deal in production with high traffic for example or heavy process that doesn't required to execute all the code process anytime, this should not be used.

I think the option 4 is the one but with a deeper explanation and some use case examples.

**Comment by zanardigit** [comment]
As per the Unix philosophy "do one thing and do it well" it makes sense to deprecate and disable HttpCache in production as there are better tools for that, so 4. is good for me.

**Comment by Toflar** [comment]
I think - as always - it very much depends on your use case. Personally, I think it's perfectly fine to use the HttpCache Proxy in production. Think of some smaller page that has a moderate number of page visits per day. Would you consider Varnish for such a use case? Probably not because the cost-benefit ratio is just not worth it (in this case `cost` is very much related to the time factor). However, those moderate number of visitors per day might still greatly benefit from the HttpCache component which is very easy to set up for developers. It's basically a one liner plus 2 lines of configuration when using the SF full stack. And because it's just PHP you don't need to setup anything else, it's really very, very easy.
 
There are so many examples for similar issues. You can quickly build a search engine using MySQL's full text search but obviously there's Elasticsearch and Co. that excel for those tasks. You can drive a huge car in a big city but it's probably easier driving a smaller one. The question is not whether to use it at all or not but it's about the use case.

I like the HttpCache but I know about its limitations. I would love to see it remain part of Symfony and I also think it would be a pity to just drop it. The limitations should "simply" get documented so that people are able to decide for themselves whether it is a good choice to use it or if they should look for different solutions.

That's my 2 cents :-)

**Comment by barryvdh** [comment]
I'm using it for some (Laravel) sites in production. I created a package for it also;  https://github.com/barryvdh/laravel-httpcache

I understand that Varnish is better,  but sometimes this is just a very easy tool to use to get faster responses with low resources on not so heavy sites. Varnish isn't worth is most of the time.. 

So I'm in favor of keeping a production mode, perhaps with clear suggestion that it isn't optimal (but still,  it does a very good job :))

**Comment by DracoBlue** [comment]
It is really nice to have this in native php and scales much better instead of no http cache at all ;).

2) is ok, because development needs to be fast an expressive.

If you would do 4) and release it only as another library next to symfony and only explain how to use it with symfony or laravel - it would be a nice addon, to optin if you really need it but with independent release cycles.

**Comment by zanardigit** [comment]
I'd like to point out, if I understand correctly, that the suggestion is not to remove **HttpCache** from Symfony, but to disable it in **production environment**. I guess nothing prevents you from creating a **custom** environment which is production-like but has HttpCache enabled also. 

Feel free to correct me if I am wrong.

**Comment by lolautruche** [comment]
HttpCache is useful for functional testing with tools like Behat. It makes it possible to quickly boot an "almost like production" environment. I know, it's not like having Varnish but it can make sense for general testing, even with known edge cases. 

Furthermore, I think it can used in production with small apps, just like @toflar says. But it must stay unrecommended and limitations should definitely be clearly documented. 
And using Varnish should be more strongly advertised IMHO.

**Comment by stof** [comment]
@crell @lolautruche Are Drupal and EzPublish using the HttpCache, and if yes, in which way ?

**Comment by andrerom** [comment]
@stof in eZ it's used *(extended with FOSHttpCache)* like in option 3, we recommend Varnish, but we have community users that can't install it and have low traffic numbers so are fine with the built in one. But unlike said above they can't be asked to run without cache, as a full app will be way to slow.

So option 3 would be the better choice from our side afaik. Removing the built in proxy will remove most of the value of http cache.

**Comment by Pierstoval** [comment]
I'm also in for recommending to use Varnish for "big projects" and HttpCache for "smaller projects".

After all, it's clearly possible to do as what've been already done for AsseticBundle: still allow people to use it, but not recommending it because of its flaws (a.k.a the "grow infinitely" HttpCache issue).

I'm also using HttpCache in small apps in production because I am no expert in Varnish and I run multiple websites/hosts on my machines in the old way, so I don't want my stack to be too heavy, but I'm using Varnish in bigger project on which I'm not the leader, and it's way more efficient than HttpCache (especially when dockerizing different applications or using many servers with a load balancer, heavy stack, etc.).

I would be ok with both options 3 and 4, but I would recommend "almost" 4 because I don't want HttpCache to be removed. Just to keep it as "available but not activated nor recommended".

Hope it helps Symfony guys to make a decision :smile:

**Comment by andrerom** [comment]
@Pierstoval Afaik it does not sound like you are pro 4, as that is removal of HttpCache. Did you mean option 2 and 3 perhaps?

**Comment by Crell** [comment]
Drupal 8 is not using HttpCache.  Rather, we're doing something very similar but with our own implementation that is more aware of Drupal's cache tags functionality.  We also recommend a for-reals reverse proxy for most installations, specifically one that supports tag-based purging (which includes Varnish).

**Comment by MasterB** [comment]
I use HttpCache at some smaller projects. For me, it works great at moment. To deprecate it makes me not really happy!
I don't need more infos in debug profiler about cache. Maybe the limitation, problems in production could  described in more detailed way. I think to spend HttpCache a little bit more love and everything is ok.

**Comment by Pierstoval** [comment]
> @Pierstoval Afaik it does not sound like you are pro 4, as that is removal of HttpCache. Did you intend to say option 2 and 3 perhaps?

Yes indeed, sorry, mis-counted proposals :smile:

**Comment by joschi127** [comment]
Still great for a) smaller projects and as well b) bigger projects with limitations what you can setup on the used server environment (e. g. we often have intranet or extranet applications hosted by the customer themselves and they only provide a standard basic LAMP setup)

**Comment by davedevelopment** [comment]
Holding my hands up, I've been running HttpCache in production for years. It started out because we were running on managed servers, where I didn't have the option to run varnish, HttpCache allowed me to implement caching in a sensible way. We've since moved on and all of our infrastructure is on AWS, but we never got around to putting some varnish in place...

**Comment by senkal** [comment]
I used to use Varnish even in dev environment to simulate better production.
Never had to use HttpCache but I see some cases when, for some people, can be useful.

Main argument: if it works, it's not stupid.
Can be not recommended but should be available as option, for me number 3 is fine.

Number 4 for me can be described as: everybody who uses it in production(I mean you make money and it works) you do it wrong, we know better how your infrastructure should look like.
No Varnish/other pre-proxy? Your problem. 
No valid use case for Varnish? Your problem.

**Comment by Toflar** [comment]
Fun fact: I'm currently trying to implement it into Contao (a CMS) and that's exactly where HttpCache might be a wonderful starting point for some basic websites that will be still way more performant when using HttpCache than when having no ESI support at all. Basically it is a huge performance gain you get out of the box. Waiting for PHP to boot as a proxy might still be way faster than requesting the database every time to e.g. dynamically create a navigation from a site structure in a CMS. If you want to optimize further you can still (and yes, you should if you can) switch to Varnish. I guess this goes a bit into the same direction as eZ seems to benefit from HttpCache.

However, there are a few things that bother me (such as missing housekeeping and missing store adapters) and I think I'd be willing to work on it. If a few people want to join I think we can build some cool stuff together.
In my opinion the SF core team should try to agree on whether HttpCache should remain part of Symfony. I fully understand the point of view e.g. @jpauli outlined. So maybe HttpCache should become something like "symfony/http-kernel-cache"? Then people using Varnish don't have to be scared by these additional php classes :-P

**Comment by fabpot** [comment]
So, the community has spoken. HttpCache is here to stay and production usage is a supported use case. Time to get back to work and see how we can improve the current production usage story.

**Comment by andrerom** [comment]
> However, there are a few things that bother me (such as missing housekeeping and missing store adapters) 

Sounds like it should be changed to use Cache component so this kind of features can be added in one place, eg expiry handling and cleanup, and already existing support for adapters. It would also open the door for HTTPCache to out of the gate support tagging, which can be accomplished easily in Varnish using BAN or more performant using xkey vmod.

> So maybe HttpCache should become something like "symfony/http-kernel-cache"? Then people using Varnish don't have to be scared by these additional php classes :-P

As for naming, I would say the component as used either way is Symfony HttpCache, while the php proxy code is Symfony [HttpCache] Proxy.

**Comment by nicolas-grekas** [comment]
I agree with @andrerom: we should use Symfony Cache as backend to benefit from its storage adapters and upcoming tagging+forking features!

**Comment by odino** [comment]
Btw there will be loads of people who can't run nginx / Apache and are wondering why the PHP guys don't provide a production-ready built-in webserver :)

This is to say that's ok to listen to the community but at some point you'll have to figure out on your own what's really a priority vs what's re-inventing the wheel.

**Comment by senkal** [comment]
@odino That is exactly what happened here.
It looks like it will be much better communicated why not to use httpcache in prod.
I think the balance is right here.

You compare two not comparable things.
1. Web server is installed, and often configured, on most host providers. Pre-proxy cache servers are not (at least from my experience)
2. Web server is essential to run website, so people are much more aware of the right way to use it, where many website don't need pre-proxy caching.
3. I haven't heard about anybody saying that php built in webserver should be used in production.
4. Built in php server was communicated very well as just dev/early stage tool, I don't think many devs were even confused where to use it and where not to use it. 
Opposite to HttpCache where even experienced developers used it in production.
5. I get your point about community but you took not the best example.

**Comment by odino** [comment]
I think the final decision was to support prod usage and try to fix the
stuff that's currently broken :)

I'd rather see the efforts focused elsewhere, but that's just my 2 cents :)

**Comment by javiereguiluz** [comment]
Closing this as "fixed" because, as Fabien pointed out, a consensus was reached. Thank you all for the discussion.
